kimchi/src/circuits/domains.rs

```diff
@@ -40,13 +42,139 @@ impl<F: FftField> EvaluationDomains<F> {
+    }
+
+    /// Changes the domain of an [Evaluations] to d2.
+    pub fn to_d2(&self, e: &Evaluations<F, Domain<F>>) -> Evaluations<F, Domain<F>> {
+        let shift = Self::shift(self, e);
+        if shift < 2 {
+            e.interpolate_by_ref().evaluate_over_domain(self.d2)
+        } else {
+            Evaluations::from_vec_and_domain(
+                e.evals.iter().step_by(shift).cloned().collect(),
+                self.d2,
+            )
+        }
+    }
+
+    /// Changes the domain of an [Evaluations] to d4.
+    pub fn to_d4(&self, e: &Evaluations<F, Domain<F>>) -> Evaluations<F, Domain<F>> {
+        let shift = Self::shift(self, e);
+        if shift < 4 {
+            e.interpolate_by_ref().evaluate_over_domain(self.d4)
+        } else {
+            Evaluations::from_vec_and_domain(
+                e.evals.iter().step_by(shift).cloned().collect(),
+                self.d2,
+            )
+        }
+    }
+
+    /// Changes the domain of an [Evaluations] to d8.
+    pub fn to_d8(&self, e: &Evaluations<F, Domain<F>>) -> Evaluations<F, Domain<F>> {
+        let shift = Self::shift(self, e);
+        if shift < 8 {
+            e.interpolate_by_ref().evaluate_over_domain(self.d8)
+        } else {
+            Evaluations::from_vec_and_domain(
+                e.evals.iter().step_by(shift).cloned().collect(),
+                self.d2,
+            )
+        }
+    }
+}
+
+/// A wrapper around [Evaluations] that implements [Add].
+/// This allows adding polynomials in evaluation forms,
+/// even though they might have different domains.
+///
+/// **Warning**: this works between the fields d1, d2, d4, and d8,
+/// created by [EvaluationDomains::create] due to a relation between the domains.
+/// This function should not be used for other domains.
+pub struct Evals<F: FftField>(pub Evaluations<F, Domain<F>>);
+
+impl<F> Deref for Evals<F>
+where
+    F: FftField,
+{
+    type Target = Evaluations<F, Domain<F>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<F> From<Evaluations<F, Domain<F>>> for Evals<F>
+where
+    F: FftField,
+{
+    fn from(evals: Evaluations<F, Domain<F>>) -> Self {
+        Self(evals)
+    }
+}
+
+impl<F> Add<Self> for &Evals<F>
+where
+    F: FftField,
+{
+    type Output = Evals<F>;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        let lhs_size = self.domain().size();
+        let rhs_size = rhs.domain().size();
+
+        let (bigger, smaller) = if lhs_size > rhs_size {
+            (self, rhs)
+        } else {
+            (rhs, self)
+        };
+
+        let bigger_size = bigger.domain().size();
+        let smaller_size = smaller.domain().size();
+
+        let step = match bigger_size {
+            x if x == smaller_size => 1,
+            x if x == smaller_size * 2 => 2,
+            x if x == smaller_size * 4 => 4,
+            x if x == smaller_size * 8 => 8,
```

Comment:
one idea to make this safer is to check here what we enforce in [create]:

```rust
x if x == smaller_size * 2 => { assert_eq!(smaller.group_gen.square(), bigger.group_gen); 2 }
```

Reply:
another way to deal with this is to pass a reference to `domains: &EvaluationDomains<F>` and enforce that the domain of both the `lhs` and `rhs` is among the ones defined in our `domains`.